FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,87 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				if let cache = try ManagedCache.find(in: context) {
+					completion(.found(
+						feed: cache.feed
+							.compactMap { ($0 as? ManagedFeedImage) }
+							.map {
+								LocalFeedImage(id: $0.id,
+								               description: $0.imageDescription,
+								               location: $0.location,
+								               url: $0.url)
+							},
+						timestamp: cache.timestamp))
```

Comment:
Suggestion:
To improve this part's readability, it can be moved to a separate local constant or a helper method as shown in the course.
```suggestion
					let feed = cache.feed
						.compactMap { ($0 as? ManagedFeedImage) }
						.map {
							LocalFeedImage(id: $0.id,
							               description: $0.imageDescription,
							               location: $0.location,
							               url: $0.url)
						}
					completion(.found(
						feed: feed,
						timestamp: cache.timestamp))
```